Index: linux-4.4.173/drivers/mtd/devices/m25p80.c
===================================================================
--- linux-4.4.173.orig/drivers/mtd/devices/m25p80.c	2019-02-26 08:28:43.049294471 +0000
+++ linux-4.4.173/drivers/mtd/devices/m25p80.c	2019-02-26 08:34:37.549559607 +0000
@@ -294,7 +294,7 @@
 	{"mx25l25635e"},{"mx66l51235l"},
 	{"n25q064"},	{"n25q128a11"},	{"n25q128a13"},	{"n25q512a"},
 	{"s25fl256s1"},	{"s25fl512s"},	{"s25sl12801"},	{"s25fl008k"},
-	{"s25fl064k"}, {"s25fl164k"},
+	{"s25fl064k"}, {"s25fl164k"}, {"s25fl256l"},
 	{"sst25vf040b"},{"sst25vf016b"},{"sst25vf032b"},{"sst25wf040"},
 	{"m25p40"},	{"m25p80"},	{"m25p16"},	{"m25p32"},
 	{"m25p64"},	{"m25p128"},
Index: linux-4.4.173/drivers/mtd/spi-nor/spi-nor.c
===================================================================
--- linux-4.4.173.orig/drivers/mtd/spi-nor/spi-nor.c	2019-02-26 08:28:42.981294437 +0000
+++ linux-4.4.173/drivers/mtd/spi-nor/spi-nor.c	2019-02-26 08:36:06.853647357 +0000
@@ -780,6 +780,7 @@
 	{ "s25fl132k",  INFO(0x014016,      0,  64 * 1024,  64, SECT_4K) },
 	{ "s25fl164k",  INFO(0x014017,      0,  64 * 1024, 128, SECT_4K) },
 	{ "s25fl204k",  INFO(0x014013,      0,  64 * 1024,   8, SECT_4K | SPI_NOR_DUAL_READ) },
+	{ "s25fl256l",  INFO(0x016019,      0,  64 * 1024, 512, SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
 
 	/* SST -- large erase sizes are "overlays", "sectors" are 4K */
 	{ "sst25vf040b", INFO(0xbf258d, 0, 64 * 1024,  8, SECT_4K | SST_WRITE) },
